runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricResults.java

```diff
@@ -0,0 +1,184 @@
+              MetricKey key = metricResult.getKey();
+              return new SparkMetricResult<>(key.metricName(), key.stepName(),
+                  metricResult.getValue());
+            }
+          });
+    }
+
+    @Override
+    public Iterable<MetricResult<DistributionResult>> distributions() {
+      Iterable<DistributionAggregator> allDistributions =
+          metricsAccum.value().getDistributions();
+      Iterable<DistributionAggregator> filteredDistributions =
+          Iterables.filter(allDistributions,
+              new Predicate<DistributionAggregator>() {
+            @SuppressWarnings("ConstantConditions")
+            @Override
+            public boolean apply(DistributionAggregator metricResult) {
+              return matches(filter, metricResult.getKey());
+            }
+          });
+      return Iterables.transform(filteredDistributions,
+          new Function<DistributionAggregator, MetricResult<DistributionResult>>() {
+            @SuppressWarnings("ConstantConditions")
+            @Nullable
+            @Override
+            public MetricResult<DistributionResult>
+            apply(DistributionAggregator metricResult) {
+              MetricKey key = metricResult.getKey();
+              return new SparkMetricResult<>(key.metricName(), key.stepName(),
+                  metricResult.getValue().extractResult());
+            }
+          });
+    }
+
+    private boolean matches(MetricsFilter filter, MetricKey key) {
+      return matchesName(key.metricName(), filter.names())
+          && matchesScope(key.stepName(), filter.steps());
+    }
+
+    private boolean matchesName(MetricName metricName, Set<MetricNameFilter> nameFilters) {
+      if (nameFilters.isEmpty()) {
+        return true;
+      }
+
+      for (MetricNameFilter nameFilter : nameFilters) {
+        if ((nameFilter.getName() == null || nameFilter.getName().equals(metricName.name()))
+            && Objects.equal(metricName.namespace(), nameFilter.getNamespace())) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+
+    private boolean matchesScope(String actualScope, Set<String> scopes) {
+      if (scopes.isEmpty() || scopes.contains(actualScope)) {
+        return true;
+      }
+
+      for (String scope : scopes) {
+        if (actualScope.startsWith(scope)) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+  }
+
+  private static class SparkMetricResult<T> implements MetricResult<T> {
+    private final MetricName name;
+    private final String step;
+    private final T result;
+
+    SparkMetricResult(MetricName name, String step, T result) {
+      this.name = name;
+      this.step = step;
+      this.result = result;
+    }
+
+    @Override
+    public MetricName name() {
+      return name;
+    }
+
+    @Override
+    public String step() {
+      return step;
+    }
+
+    @Override
+    public T committed() {
+      return result;
```

Comment:
Is the result the committed value or the attempted value? It may be better to throw an UnsupportedOperationException on whichever type of value the result isn't to make it clear what semantics the value has.